tools/reaper/internal/internal.go

```diff
@@ -86,82 +95,59 @@ func Serve(conn net.Listener, params *Params) error {
-	resp, err := c.ListClusters(ctx, &containerpb.ListClustersRequest{
-		Parent: fmt.Sprintf("projects/%s/locations/%s", params.ProjectID, params.Location),
-	})
+	namespaceInterface := kubeClient.CoreV1().Namespaces()
+	namespaces, err := namespaceInterface.List(metav1.ListOptions{})
 	if err != nil {
-		return []*containerpb.Cluster{}, err
+		return errors.Wrap(err, "listing namespace from kubernetes cluster failed")
 	}
-	orphanedClusters := []*containerpb.Cluster{}
-	for _, cluster := range resp.Clusters {
-		if isOrphaned(cluster, params) {
-			log.Printf("Cluster '/%s' was orphaned.\nDetails: %+v", cluster.Name, cluster)
-			orphanedClusters = append(orphanedClusters, cluster)
-		}
-	}
-	return orphanedClusters, nil
-}
 
-func deleteClusters(ctx context.Context, c *container.ClusterManagerClient, clusters []*containerpb.Cluster) error {
-	for _, cluster := range clusters {
-		clusterName := selfLinkToQualifiedName(cluster.SelfLink)
-		log.Printf("Deleting Orphaned GKE Cluster %s", clusterName)
-		resp, err := c.DeleteCluster(ctx, &containerpb.DeleteClusterRequest{
-			Name: clusterName,
-		})
-		if err != nil {
-			return err
+	for _, namespace := range namespaces.Items {
+		if isOrphaned(namespace, params) {
+			if err = namespaceInterface.Delete(namespace.ObjectMeta.Name, &metav1.DeleteOptions{}); err != nil {
+				// Log but dont panic the error
+				log.Printf("Error: deleting namespace %s created at %s", namespace.ObjectMeta.Name, namespace.ObjectMeta.CreationTimestamp)
+			} else {
+				log.Printf("deleting namespace %s created at %s", namespace.ObjectMeta.Name, namespace.ObjectMeta.CreationTimestamp)
+			}
+		} else {
+			log.Printf("skipping namespace %s created at %s", namespace.ObjectMeta.Name, namespace.ObjectMeta.CreationTimestamp)
 		}
-		log.Printf("Delete GKE Cluster: %s => %v", clusterName, resp)
 	}
-	return nil
+
+	return err
 }
 
-func isOrphaned(cluster *containerpb.Cluster, params *Params) bool {
-	deadline := time.Now().Add(-1 * params.Age)
-	creationTime, err := time.Parse(time.RFC3339, cluster.CreateTime)
-	if err != nil {
-		log.Printf("cannot parse time '%s' because '%v', ignoring for orphan detection.", cluster.CreateTime, err)
-		return false
-	}
-	if creationTime.After(deadline) {
-		return false
-	}
-	if _, ok := cluster.ResourceLabels[params.Label]; !ok {
-		return false
-	}
-	return true
+func fileExists(name string) bool {
+	_, err := os.Stat(name)
+	return err == nil
 }
 
-func selfLinkToQualifiedName(selfLink string) string {
-	parts := strings.Split(selfLink, "v1/")
-	if len(parts) != 2 {
-		log.Fatalf("SelfLink: %s is not a valid selflink it should contain 'v1/'", selfLink)
-	}
-	return parts[1]
+func isOrphaned(namespace corev1.Namespace, params *Params) bool {
+	deadline := time.Now().Add(-1 * params.Age).UTC()
+	return namespace.ObjectMeta.CreationTimestamp.Time.Before(deadline) && namespace.ObjectMeta.Name != "kube-system"
```

Comment:
Can we do a whitelist here instead of blacklist?  It'd be safer to look for namespaces which match the expected pattern that exclude those which don't.